### ISSUE TITLE
"rindex"/ "rfind" tool

### ISSUE THREAD

**Issue by pylang** [erikrose/more-itertools#205]
From [rhettinger's challenge][3], should we consider adding a tool like  [`rindex`][1] or [`rfind`][0] to this library?  In particular, these tools find index for the last occurrence of an item.

**Sample Implementations**

[rhettinger's solution][5]
```python
def rindex(seq, x):
    return next(compress(count(len(s)-1, -1), (map(x.__eq__, reversed(s)))), -1)
```

more-itertools solution
```python
import more_itertools as mit


def rindex(iterable, item):
    """Return the highest index for the given item."""
    return next(mit.islice_extended(mit.locate(iterable, lambda x: x == item), -1, None))
```

**Uses**

- [How to find the last occurrence of an item in a Python list][7]
- [Equivelant to rindex for lists in Python [duplicate]][8]
- [Find index of last occurrence of a substring in a string][9]
- See N. Coglan's response to [this tweet][6]

  [0]: https://docs.python.org/2/library/string.html#string.rfind
  [1]: https://docs.python.org/2/library/string.html#string.rindex
  [3]: https://twitter.com/raymondh/status/987735142016864256
  [5]: https://twitter.com/raymondh/status/987835455046479872
  [6]: https://twitter.com/Persaik/status/987771879866892288
  [7]: https://stackoverflow.com/questions/6890170/how-to-find-the-last-occurrence-of-an-item-in-a-python-list
  [8]: https://stackoverflow.com/questions/9836425/equivelant-to-rindex-for-lists-in-python
  [9]: https://stackoverflow.com/questions/9572490/find-index-of-last-occurrence-of-a-substring-in-a-string?rq=1

**Comment by bbayles** [comment]
I saw that Twitter message - nice solution!

I think the nice thing about `locate()` is that it returns all the indexes rather than just one (that was actually [your idea](https://github.com/erikrose/more-itertools/issues/139#issuecomment-302894121)!). So how about an `rlocate()`, which is basically a reversed `locate()`? I'd accept that PR.

This is close to what you have above, of course.

**Comment by pylang** [comment]
Many thanks.  Hmm.  It's true `locate()` is a bit more general by returning all indicies.  It may even obviate any of these tools `r...` tools.  It presented the idea mostly for pondering.  

`rlocate()` sounds like a good abstraction and good complement.   Let me think about that a bit.  Why we have these kinds of methods suggests there's a deeper thing going on, likely related to reversing generators.  Solving that problem efficiently (if at all possible)  would have a big impact.

**Comment by bbayles** [comment]
Re: reversing things, you can't really do much better than "store all the items and then emit them in reverse." `islice_extended()` with a negative `step` argument takes care to not cache more at a time than is needed (see [here](https://github.com/erikrose/more-itertools/pull/146#issue-123858089) for how to work this out).

I think there are two basic approaches for an `rlocate(iterable, pred)`. The first is `return list(reversed(locate(iterable, pred)))`. This has the advantage of not caching all the items in `iterable`, just the ones where `pred` is `True`. It has the disadvantage of searching from the left, which isn't good when `pred` is expensive.

The second is something like: see if the iterable has an efficient reversal method and an available length. If so, use those and search from the right. If not, cache all the items, compute the length, and search from the right.

**Comment by pylang** [comment]
Yeah I know that's that state of reversing now.  I just feel there's some trick waiting to be uncovered.  Someday.

As for `rlocate`, the more I think about this, I wonder if `reversed(list(locate(...)))` isn't enough.  A separate function may not be needed.  Otherwise maybe give it more clear name, e.g. `reversed_indices()` (but nicer).

**Comment by jferard** [comment]
Hi! I just read this discussion and something bothers me: what to do with infinite iterables (e.g. `itertools.count`)? Consider `locate(infinite_iterable, lambda x: x == n)` and `rlocate(infinite_iterable, lambda x: x == n)`. They both will be (theoretically) infinite, by design of `locate`/`rlocate`. But there is a huge difference. In the first, we expected to get new values in a reasonable time (it's the user's responsibility to give an iterable and a predicate that will make that true). But in the second case, by definition, we won't ever get any result. Thus, the case of infinite iterables should be out the the scope of `rlocate` but not of `locate`.

That said, @bbayles's sketch may be implemented as something like that (python 3.6):

    import collections.abc as ca

    def rlocate(iterable, pred=bool, expensive_pred=False):
        if (expensive_pred and isinstance(iterable, ca.Sequence) # we prefer to (and can) reverse the iterable
                or isinstance(iterable, ca.Reversible)): # reversed is (allegedly) efficient
            return compress(count(len(iterable)-1, -1), map(pred, reversed(iterable))) # exact symmetric of locate
        else:
            return reversed(list(locate(iterable, pred))) # fall through
